### PR TITLE
Change version bounds for directory and primitive

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

**Comment by harendra-kumar** [comment]
Add another commit to move `primitive` and `exceptions` under the ghc boot library comment.

**Comment by harendra-kumar** [comment]
Also, now that we have tested with `ghc` package we can remove that commit.

File: src/Streamly/Internal/Data/SmallArray.hs [hunk 1]
```diff
@@ -50,7 +50,7 @@ import Control.Monad (when)
 import Control.Monad.IO.Class (MonadIO, liftIO)
 import GHC.IO (unsafePerformIO)
 import Data.Functor.Identity (runIdentity)
-import Data.Primitive.SmallArray
+import Streamly.Internal.Data.SmallArray.Types
```
Comment (pranaysashank) [inline src/Streamly/Internal/Data/SmallArray.hs]:
```suggestion

import Streamly.Internal.Data.SmallArray.Types
```

File: src/Streamly/Internal/Data/Prim/Array.hs [hunk 2]
```diff
@@ -60,9 +60,7 @@ import Control.Monad (when)
 import Control.Monad.IO.Class (liftIO, MonadIO)
 import GHC.IO (unsafePerformIO)
 import Data.Primitive.Types (Prim(..))
-
-import Data.Primitive.PrimArray
-
+import Streamly.Internal.Data.Prim.Array.Types
```
Comment (pranaysashank) [inline src/Streamly/Internal/Data/Prim/Array.hs]:
```suggestion

import Streamly.Internal.Data.Prim.Array.Types
```

File: streamly.cabal [hunk 3]
```diff
@@ -364,14 +366,15 @@ library
                     -- constraints of these libraries should match with
                     -- the GHC versions we support
                        base              >= 4.8   &&  < 5
+                     , ghc
```
Comment (pranaysashank) [inline streamly.cabal]:
```suggestion
```

File: streamly.cabal [hunk 4]
```diff
@@ -364,14 +366,15 @@ library
...
                      , containers        >= 0.5   && < 0.7
                      , deepseq           >= 1.4.1 && < 1.5
-                     , directory         >= 1.3   && < 1.4
+                     , directory         >= 1.2.2 && < 1.4
                      , ghc-prim          >= 0.2   && < 0.6
                      , transformers      >= 0.4   && < 0.6
 
                      , heaps             >= 0.3   && < 0.4
-                     , primitive         >= 0.6.4.0   && < 0.8
+                     , primitive         >= 0.5.4.0   && < 0.8
```
Comment (pranaysashank) [inline streamly.cabal]: Align the `&&`s

File: src/Streamly/Internal/Data/SmallArray/Types.hs [hunk 5]
```diff
@@ -0,0 +1,977 @@
...
+
+#if MIN_VERSION_base(4,9,0) || MIN_VERSION_transformers(0,4,0)
+import Data.Functor.Classes (Eq1(..),Ord1(..),Show1(..),Read1(..))
+#endif
+
+#if HAVE_SMALL_ARRAY
+data SmallArray a = SmallArray (SmallArray# a)
+  deriving Typeable
+#else
+newtype SmallArray a = SmallArray (Array a) deriving
```
Comment (pranaysashank) [inline src/Streamly/Internal/Data/SmallArray/Types.hs]: We will have SmallArray available in all the versions we support. There is no need to have this newtype wrapper over array.

File: src/Streamly/Internal/Data/SmallArray/Types.hs [hunk 6]
```diff
@@ -0,0 +1,977 @@
...
+  , sizeofSmallArray
+  , sizeofSmallMutableArray
+  , smallArrayFromList
+  , smallArrayFromListN
+  , mapSmallArray'
+  , traverseSmallArrayP
+  ) where
+
+
+#if (__GLASGOW_HASKELL__ >= 710)
```
Comment (pranaysashank) [inline src/Streamly/Internal/Data/SmallArray/Types.hs]: We always have small array.

File: src/Streamly/Internal/Data/SmallArray/Types.hs [hunk 7]
```diff
@@ -0,0 +1,977 @@
...
+-- entire array. These advantages make them suitable for use as arrays that are
+-- known to be small.
+--
+-- The card size is 128, so for uses much larger than that, 'Array' would likely
+-- be superior.
+--
+-- The underlying type, 'SmallArray#', was introduced in GHC 7.10, so prior to
+-- that version, this module simply implements small arrays as 'Array'.
+
+module Streamly.Internal.Data.SmallArray.Types
```
Comment (pranaysashank) [inline src/Streamly/Internal/Data/SmallArray/Types.hs]: Remove all the newtype wrappers in the module.

File: streamly.cabal [hunk 8]
```diff
@@ -307,7 +307,9 @@ library
                      , Streamly.Internal.Data.Time.Clock
                      , Streamly.Internal.Data.SVar
                      , Streamly.Internal.Data.Array
+                     , Streamly.Internal.Data.Prim.Array.Types
                      , Streamly.Internal.Data.Prim.Array
+                     , Streamly.Internal.Data.SmallArray.Types
```
Comment (pranaysashank) [inline streamly.cabal]: No need to export for GHC versions that come with a primitive version that contains SmallArray

File: src/Streamly/Internal/Data/SmallArray.hs (hunk 1 above)
Comment (pranaysashank) [reply src/Streamly/Internal/Data/SmallArray.hs]: We can import from Data.Primitive.SmallArray if we are using a GHC / primitive package that comes with SmallArray and save on compilation time.

**Comment by pranaysashank** [comment]
> Add another commit to move `primitive` and `exceptions` under the ghc boot library comment. `mtl` as well.

exceptions and primitive are no longer core libraries under recent ghc versions

File: src/Streamly/Internal/Data/SmallArray.hs (hunk 1 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/SmallArray.hs]: That is what I have written in the beginning but as @harendra-kumar mentioned, debugging arrays in the future might prove difficult among other problems.

Comment (pranaysashank) [reply src/Streamly/Internal/Data/SmallArray.hs]: No point in compiling everything twice!

File: streamly.cabal (hunk 3 above)
Comment (adithyaov) [reply streamly.cabal]: Will delete the commit containing this change in the end. There is still a little to be done.

File: src/Streamly/Internal/Data/SmallArray/Types.hs (hunk 5 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/SmallArray/Types.hs]: We will not be using `SmallArray` from `primitive`.

Comment (pranaysashank) [reply src/Streamly/Internal/Data/SmallArray/Types.hs]: What do you mean?

File: streamly.cabal (hunk 8 above)
Comment (adithyaov) [reply streamly.cabal]: That was the initial plan but We will not be using `SmallArray` from `primitive`.

File: src/Streamly/Internal/Data/SmallArray/Types.hs (hunk 5 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/SmallArray/Types.hs]: We will our version of `SmallArray`.

Comment (pranaysashank) [reply src/Streamly/Internal/Data/SmallArray/Types.hs]: I meant we will have SmallArray# in all the versions we support. No need to use the newtype wrapper and Array based SmallArray

File: src/Streamly/Internal/Data/SmallArray.hs (hunk 1 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/SmallArray.hs]: We might want to rename `SmallArray` to `Array` in the future. We will not be using `Data.Primitive.SmallArray` even if it is available.

File: src/Streamly/Internal/Data/SmallArray/Types.hs (hunk 6 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/SmallArray/Types.hs]: We always have HAVE_SMALL_ARRAY*

File: streamly.cabal [hunk 9]
```diff
@@ -616,6 +618,17 @@ test-suite parallel-loops
     , base   >= 4.8   && < 5
     , random >= 1.0.0 && < 2
 
+test-suite version-bounds
```
Comment (adithyaov) [inline streamly.cabal]: We can remove this test-suite once we add fusion-plugin as a dependency

**Comment by harendra-kumar** [comment]
Merged. Please also update the credits/license for primitive code used in this PR.

**Comment by adithyaov** [comment]
Already merged.

**Pull request closed without merging**